### ISSUE TITLE
Remember past searches and watch progress.

### ISSUE THREAD

**Issue by NBTX** [ApolloTVofficial/kamino#55]
**What feature would you like added to the app?**
The app should store watch progress for content and remember past searches.

**Describe the solution you'd like (optional):**
- Watch progress should be stored for content that has been watched after a certain point (i.e. not stored before 10 seconds has been watched).
    - If watch progress is available, a dialog should be displayed: "_You've already started watching this. Do you want to continue from --:-- or start again?_"
- Search history should be displayed underneath the search box by default when making a search and should remain for as long as the recently searched term starts with the search box content.

**Additional context**
![image](https://user-images.githubusercontent.com/43181178/50368316-3e8e2800-057f-11e9-8345-d15d03fb9056.png)

**Comment by edmundackah** [comment]
have implemented search history will take a look at watch progress at a later date.

**Comment by SpajicM** [comment]
Watch progress could be a problem, because sources have different length files.
I'd say most common use case scenario with this isn't "I will watch the last 10minutes tomorrow" but rather "This source is lagging, let me switch to a lower resoultion"
But that other source may have "Netflix intro" or "Previously in..." segment which the previous source didn't have, so watch progress would not get you back to a correct place.

**Comment by SpajicM** [comment]
@NBTX Sure, but as I said, if I quit watching a movie mid-way, it's usually to change source :)

**Comment by NBTX** [comment]
That's true. I suppose we could allow users to jump to where they left off regardless of the source being different. (We could maybe add to the resume message that the source length is different) 

I feel like you could probably just skip back/forwards until you have the right place.

**Comment by kevincosmos** [comment]
Renaming issue as watch history has been implemented

**Comment by polishboi** [comment]
I know this may be overkill, more important things require attention, but one way to ensure that the resume always starts at the same place regardless of a different source is to in addition of saving the timestamp, take a screenshot. Then once in the new source, recursively start from the saved timestamp checking whether the current frame is same as the screenshot: check if the previous frame is the same as screenshot, if not check next frame, if not check the second previous frame, second next etc.

**Comment by SpajicM** [comment]
@polishboi Easier said than done :)

But this seemed like a fun problem so I've decided to look into it.

What's needed here is **perceptual image hashing**  - generating a fingerprint based on what image "looks like" overall.

Three algorithms seem most common, aHash, dHash and pHash, and accuracy to performance ratio would look something like this:
![image](https://user-images.githubusercontent.com/21988220/54789008-6b5ee380-4c31-11e9-9239-865e2f5003b5.png)

I've looked around GitHub, but didn't find anyone did any image hashing in Flutter or Dart.
However, I've found a great [blog post](https://jenssegers.com/61/perceptual-image-hashes) explaining the steps for each of them, so I implemented aHash and dHash.
First, I needed an image manipulation library - https://pub.dartlang.org/packages/image seems to be an only one for Dart but it has all I needed!
The downside of it is that it's not really optimized for Flutter or mobile devices, so it's slow (https://github.com/brendan-duncan/image/issues/72), but the author of the library is working on it, and it may be good enough for non-heavy images (frames from the video, compressed).

I've published a project sample here: https://github.com/SpajicM/flutter_image_hashing
[Screenshot 1](https://user-images.githubusercontent.com/21988220/54789602-571be600-4c33-11e9-866e-148305336cf0.png) | [Screenshot 2](https://user-images.githubusercontent.com/21988220/54789603-571be600-4c33-11e9-932c-bc58f7ff1e61.png)

The next step would be to get the video frame we want to get hash for. There are 2 approaches to this:
1) Getting a frame image from a video player
2) Taking a screenshot of a video player widget.

The "official" video player plugin for Flutter (https://pub.dartlang.org/packages/video_player) doesn't have the feature of capturing a video frame directly. As for the second approach, Flutter supports taking a screenshot using [RenderRepaintBoundary](https://master-docs-flutter-io.firebaseapp.com/flutter/rendering/RenderRepaintBoundary/toImage.html) but Flutter's video player is a plugin, using platform's native players in the background, ExoPlayer (Android) and AVPlayer (iOS) which makes it not-technically-a-widget. I've tried it, and it only was able to capture video player controls, but not the video itself.
Only approach to this for now would be extending Flutter video player plugin with the native methods for both platform, MediaMetadataRetriever for Android and AVAssetImageGenerator for iOS.

We may do this in the future, if image processing library gets improved for Flutter and if practice shows length/cut difference in cuts between link sources is common. Till then, you may have to manually seek backward/forward a bit.

**Comment by NBTX** [comment]
@SpajicM That's pretty awesome and I like that you worked with the screenshot idea - I would have just dismissed it without thinking of using a distance/difference algorithm.

Anyway I just wanted to say that it might be possible - though I don't know how slow it actually was - because we could possible cut down the range of frames that we have to search. I guess you used dhash just from glancing but it may also be possible to simply use ahash instead.

**Comment by Chaphasilor** [comment]
Until the visual pinpointing is implemented, you could just always start a minute or 30 seconds earlier. If you are continuing to watch the movie/episode the next day you probably won't mind re-watching the previous 30 seconds, and if you are switching sources you might have to skip ahead a few times, given you mind re-watching a few seconds.

I find the visual approach fascinating, but if I'm not mistaken this would only work with the default player, right?
